App/Services/UriHandler.cs

```diff
@@ -0,0 +1,150 @@
+using System;
+using System.Collections.Specialized;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+using System.Web;
+using Coder.Desktop.App.Models;
+using Coder.Desktop.Vpn.Proto;
+using Microsoft.Extensions.Logging;
+
+
+namespace Coder.Desktop.App.Services;
+
+public interface IUriHandler : IAsyncDisposable
+{
+    public Task HandleUri(Uri uri, CancellationToken ct = default);
+}
+
+public class UriHandler(
+    ILogger<UriHandler> logger,
+    IRpcController rpcController,
+    IUserNotifier userNotifier,
+    IRdpConnector rdpConnector) : IUriHandler
+{
+    private const string OpenWorkspacePrefix = "/v0/open/ws/";
+
+    internal class UriException(string title, string detail) : Exception
```

Comment:
AFAIK this creates an exception with no message. It's fine for your use case of showing a notification only, but it probably should still be a valid exception.

You'll need to create a proper constructor function and call the base constructor.

e.g.

```cs
internal class UriException : Exception
{
  // ...
  internal UriException(string title, string detail) : base($"{title}: {detail}")
  {
    Title = title;
    Detail = detail;
  }
}
```